File: pine-python/pine/cli/cause_chain_probe.py
"""pine-python cause-chain-probe: smoke test for ExecutionError cause-chain
unwrap.

Constructs a FakeRedisError, wraps it via ``raise ... from`` into
pine.errors.ExecutionError, then uses ``isinstance(err.__cause__, T)`` to
recover the inner cause. Prints either::

    PASS:<recovered_inner_msg>

or::

    FAIL:<reason>

on stdout. cross-validate Section 15 asserts byte-identical stdout across
pine-go / pine-java / pine-python / pine-cpp probes.
"""
from __future__ import annotations

import sys

from pine.errors import ExecutionError


class FakeRedisError(Exception):
    def __init__(self, key: str):
        super().__init__(f"key={key} not found")
        self.key = key


def main() -> int:
    try:
        try:
            raise FakeRedisError("user:42")
        except FakeRedisError as inner:
            raise ExecutionError("redis_getter", inner) from inner
    except ExecutionError as outer:
        if isinstance(outer.__cause__, FakeRedisError):
            print(f"PASS:{outer.__cause__}")
            return 0
        print("FAIL:__cause__ did not recover FakeRedisError from ExecutionError chain")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: pine-java/src/main/java/page/liam/pine/CauseChainProbe.java
package page.liam.pine;

/**
 * Smoke test for ExecutionError cause-chain unwrap.
 *
 * Constructs a FakeRedisError, wraps it via {@code new ExecutionError(op,
 * cause)} into pine ExecutionError, then uses {@code getCause() instanceof
 * FakeRedisError} to recover the inner cause. Prints either:
 *
 * <pre>
 *   PASS:&lt;recovered_inner_msg&gt;
 * </pre>
 *
 * or:
 *
 * <pre>
 *   FAIL:&lt;reason&gt;
 * </pre>
 *
 * on stdout. cross-validate Section 15 asserts byte-identical stdout across
 * pine-go / pine-java / pine-python / pine-cpp probes.
 */
public final class CauseChainProbe {
    static final class FakeRedisError extends RuntimeException {
        FakeRedisError(String key) {
            super("key=" + key + " not found");
        }
    }

    public static void main(String[] args) {
        try {
            try {
                throw new FakeRedisError("user:42");
            } catch (FakeRedisError inner) {
                throw new PineErrors.ExecutionError("redis_getter", inner);
            }
        } catch (PineErrors.ExecutionError outer) {
            Throwable cause = outer.getCause();
            if (cause instanceof FakeRedisError) {
                System.out.println("PASS:" + cause.getMessage());
                return;
            }
            System.out.println("FAIL:getCause() did not recover FakeRedisError from ExecutionError chain");
            System.exit(1);
        }
    }
}
